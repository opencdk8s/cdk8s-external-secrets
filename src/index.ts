import { ApiObject, GroupVersionKind } from 'cdk8s';
import { Construct } from 'constructs';

import * as k8s from './imports/k8s';
export * as k8s from './imports/k8s';

export interface SecretStoreRef {
  readonly name: string;
  readonly kind?: string;
}

export interface ExternalSecretTemplateMetadata {
  readonly annotations?: { [key: string]: string};
  readonly labels?: { [key: string]: string};
}

export interface ExternalSecretTemplate {
  readonly type?: string;
  readonly engineVersion?: string;
  readonly metadata?: ExternalSecretTemplateMetadata;
  readonly data?: { [key: string]: string};
  readonly templateFrom?: TemplateFrom[];
}

export interface TemplateFrom {
  readonly configMap?: TemplateRef;
  readonly secret?: TemplateRef;
}

export interface TemplateRef {
  readonly name: string;
  readonly items: TemplateRefItem[];
}

export interface TemplateRefItem {
  readonly key: string;
}

export interface ExternalSecretTarget {
  readonly name?: string;
  readonly creationPolicy?: string;
  readonly deletionPolicy?: string;
  readonly template?: ExternalSecretTemplate;
  readonly immutable?: boolean;
}

export interface ExternalSecretData {
  readonly secretKey: string;
  readonly remoteRef: ExternalSecretDataRemoteRef;
}

export interface ExternalSecretDataRemoteRef {
  readonly key: string;
  readonly metadataPolicy?: string;
  readonly property?: string;
  readonly version?: string;
  readonly conversionStrategy?: string;
  readonly decodingStrategy?: string;
}

export interface ExternalSecretDataFromRemoteRef {
  readonly extract?: ExternalSecretDataRemoteRef;
  readonly find?: ExternalSecretFind;
  readonly rewrite?: ExternalSecretRewrite[];
}

export interface ExternalSecretRewrite {
  readonly regexp?: ExternalSecretRewriteRegexp;
}

export interface ExternalSecretRewriteRegexp {
  readonly source: string;
  readonly target: string;
}

export interface ExternalSecretFind {
  readonly path?: string;
  readonly name?: FindName;
  readonly tags?: { [key: string]: string};
  readonly conversionStrategy?: string;
  readonly decodingStrategy?: string;
}

export interface FindName {
  readonly regexp?: string;
}

export interface ExternalSecretSpec {
  readonly secretStoreRef: SecretStoreRef;
  readonly target?: ExternalSecretTarget;
  readonly refreshInterval?: string;
  readonly data?: ExternalSecretData[];
  readonly dataFrom?: ExternalSecretDataFromRemoteRef[];
}

export interface ExternalSecretProps {
  readonly metadata: k8s.ObjectMeta;
  readonly spec: ExternalSecretSpec;
}

export class ExternalSecretV1Beta1 extends ApiObject {
  public static readonly GVK: GroupVersionKind = {
    apiVersion: 'external-secrets.io/v1beta1',
    kind: 'ExternalSecret',
  };
  /**
   * @param props initialization props
   */
  public static manifest(props: ExternalSecretProps): any {
    return {
      ...ExternalSecretV1Beta1.GVK,
      ...props,
    };
  }

  /**
   * @param scope the scope in which to define this object
   * @param id a scope-local name for the object
   * @param props initialization props
   */
  public constructor(scope: Construct, id: string, props: ExternalSecretProps) {
    super(scope, id, ExternalSecretV1Beta1.manifest(props));
  }
}
