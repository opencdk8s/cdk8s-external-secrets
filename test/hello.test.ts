import { Testing, Chart } from 'cdk8s';
import * as t from '../src/index';
test('hello', () => {
  const app = Testing.app();
  const chart = new Chart(app, 'test');
  new t.ExternalSecretV1Beta1(chart, 'test', {
    metadata: {
      name: 'test',
    },
    spec: {
      refreshInterval: '1h',
      secretStoreRef: {
        kind: 'SecretStore',
        name: 'example',
      },
      target: {
        name: 'secret-to-be-created',
        creationPolicy: 'Owner',
      },
      dataFrom: [
        {
          extract: {
            key: 'all-keys-example-secret',
          },
        },
      ],
    },
  });


  expect(Testing.synth(chart)).toMatchSnapshot();
});
