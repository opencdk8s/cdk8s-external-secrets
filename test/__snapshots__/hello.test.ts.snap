// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello 1`] = `
Array [
  Object {
    "apiVersion": "external-secrets.io/v1beta1",
    "kind": "SecretProviderClass",
    "metadata": Object {
      "name": "test",
    },
    "spec": Object {
      "dataFrom": Array [
        Object {
          "extract": Object {
            "key": "all-keys-example-secret",
          },
        },
      ],
      "refreshInterval": "1h",
      "secretStoreRef": Object {
        "kind": "SecretStore",
        "name": "example",
      },
      "target": Object {
        "creationPolicy": "Owner",
        "name": "secret-to-be-created",
      },
    },
  },
]
`;
